crates/evm/core/src/backend/mod.rs

```diff
@@ -1939,67 +1941,81 @@ fn merge_journaled_state_data(
+        trace!("merging contract cache");
+        fork_db.contracts.insert(acc.info.code_hash, code.clone());
+    }
+
+    // port account storage over
+    use std::collections::hash_map::Entry;
+    match fork_db.accounts.entry(addr) {
+        Entry::Vacant(vacant) => {
+            trace!("target account not present - inserting from active");
+            // if the fork_db doesn't have the target account
+            // insert the entire thing
+            vacant.insert(acc.clone());
+        }
+        Entry::Occupied(mut occupied) => {
+            trace!("target account present - merging storage slots");
+            // if the fork_db does have the system,
+            // extend the existing storage (overriding)
+            let fork_account = occupied.get_mut();
+            fork_account.storage.extend(&acc.storage);
+        }
     }
-
-    fork_db.accounts.insert(addr, acc);
 }
 
 /// Clones the zk account data from the `active` db into the `ForkDB`
+#[tracing::instrument(level = "trace", skip(active, fork_db))]
 fn merge_zk_account_data<ExtDB: DatabaseRef>(
     addr: Address,
     active: &CacheDB<ExtDB>,
     fork_db: &mut ForkDB,
 ) {
-    trace!(?addr, "merging zk database data");
-
-    // TODO: do the same for nonce and codes
-    let balance_addr = L2_BASE_TOKEN_ADDRESS.to_address();
-    let mut acc = if let Some(acc) = active.accounts.get(&balance_addr).cloned() {
-        acc
-    } else {
-        // Account does not exist
-        return;
+    let mut merge_entry = |system_contract: Address, slot: U256| {
+        let Some(active_system) = active.accounts.get(&system_contract) else { return };
+        let Some(active_slot_value) = active_system.storage.get(&slot) else { return };
+
+        use foundry_zksync_core::convert::ConvertRU256;
+        tracing::trace_span!("merge_entry", ?system_contract, slot = ?slot.to_h256(), value = ?active_slot_value.to_h256());
+
+        use std::collections::hash_map::Entry;
+        match fork_db.accounts.entry(system_contract) {
+            Entry::Vacant(vacant) => {
+                trace!("system contract not present - inserting from active");
+                // if the fork_db doesn't have system,
+                // import the active one, and only populate the slot we are processing
+                let imported_system = revm::db::DbAccount {
+                    info: active_system.info.clone(),
+                    account_state: active_system.account_state.clone(),
+                    storage: [(slot, *active_slot_value)].into_iter().collect(),
+                };
+                vacant.insert(imported_system);
+            }
+            Entry::Occupied(mut occupied) => {
+                trace!("system contract present - merging slot");
+                // if the fork_db does have the system,
+                // only override the target slot
+                occupied.get_mut().storage.insert(slot, *active_slot_value);
```

Comment:
Similar to above this was mirroring the original foundry code and if possible we should keep it like that to help incorporate any future changes that might occur. Though we can have optimized code for zk specific changes